### PR TITLE
Expose the orbit analyser through Vessel and the interface

### PR DESCRIPTION
**Pull request by eggrobin**
This is the C++ backing for #2337.

### CODE REVIEW

File: ksp_plugin/interface_body.hpp [hunk 1]
```diff
@@ -167,6 +172,53 @@ inline bool operator==(NavigationManoeuvreFrenetTrihedron const& left,
...
+                       EquatorialCrossings const& right) {
+  return left.longitudes_reduced_to_ascending_pass ==
+             right.longitudes_reduced_to_ascending_pass &&
+         left.longitudes_reduced_to_descending_pass ==
+             right.longitudes_reduced_to_descending_pass;
+}
+
+inline bool operator==(OrbitGroundTrack const& left,
+                       OrbitGroundTrack const& right) {
+  return left.equatorial_crossings ==
```
Comment (pleroy) [inline ksp_plugin/interface_body.hpp]: Would fit on one line.

File: serialization/journal.proto [hunk 2]
```diff
@@ -155,8 +155,65 @@ message XY {
   required double y = 2;
 }
 
+message Interval {
```
Comment (pleroy) [inline serialization/journal.proto]: Document the ordering like is done at line 13-14.

File: serialization/journal.proto [hunk 3]
```diff
@@ -155,8 +155,65 @@ message XY {
...
+  required double nodal_precession = 4;
+  required Interval mean_semimajor_axis = 5;
+  required Interval mean_eccentricity = 6;
+  required Interval mean_inclination = 7;
+  required Interval mean_longitude_of_ascending_nodes = 8;
+  required Interval mean_argument_of_periapsis = 9;
+}
+
+message OrbitRecurrence {
+  // νₒ.
```
Comment (pleroy) [inline serialization/journal.proto]: Comment at end of line?

File: serialization/journal.proto [hunk 4]
```diff
@@ -155,8 +155,65 @@ message XY {
...
+  required EquatorialCrossings equatorial_crossings = 1;
+  // TODO(egg): mean solar times of the nodes.
+}
+
+message OrbitAnalysis {
+  required double progress_of_next_analysis = 1;
+  required int32 primary_index = 2;
+  required double mission_duration = 3;
+  // NOTE(egg): we currently have no support in the interface generator for
+  // optional non-string fields in interchange structs.  We use separate
```
Comment (pleroy) [inline serialization/journal.proto]: Why don't we fix the generator instead of doing this horrible hack?

File: ksp_plugin/vessel.hpp [hunk 5]
```diff
@@ -173,6 +176,11 @@ class Vessel {
       PileUp::PileUpForSerializationIndex const&
           pile_up_for_serialization_index);
 
+  void RefreshOrbitAnalysis(not_null<RotatingBody<Barycentric> const*> primary,
+                            Time const& mission_duration);
+  double progress_of_orbit_analysis() const;
```
Comment (pleroy) [inline ksp_plugin/vessel.hpp]: Blank line before.

File: serialization/journal.proto (hunk 4 above)
Comment (eggrobin) [reply serialization/journal.proto]: Fixing the generator is exceedingly complicated; we cannot have an `std::optional<T>` on the C++ side as it does not have standard layout, so we would need to define our own optional type. On the C# side, we would need to marshal the `T?` completely by hand.

**Pull request merged**